packages/jest-runtime/src/index.ts

```diff
@@ -1263,18 +1265,35 @@ class Runtime {
     resolve.paths = (moduleName: string) =>
       this._requireResolvePaths(from.filename, moduleName);
 
+    const moduleRegistry = this._moduleRegistry;
     const moduleRequire = (options && options.isInternalModule
       ? (moduleName: string) =>
           this.requireInternalModule(from.filename, moduleName)
       : this.requireModuleOrMock.bind(
           this,
           from.filename,
         )) as LocalModuleRequire;
-    moduleRequire.cache = Object.create(null);
     moduleRequire.extensions = Object.create(null);
     moduleRequire.requireActual = this.requireActual.bind(this, from.filename);
     moduleRequire.requireMock = this.requireMock.bind(this, from.filename);
     moduleRequire.resolve = resolve;
+    moduleRequire.cache = (target => {
+      const notPermittedMethod = () => {
+        console.warn('`require.cache` modification is not permitted');
+        return true;
+      };
+      return new Proxy(target, {
+        defineProperty: notPermittedMethod,
+        deleteProperty: notPermittedMethod,
+        get(_target, key) {
+          return typeof key === 'string' ? moduleRegistry.get(key) : undefined;
+        },
+        has(_target, key) {
+          return typeof key === 'string' && moduleRegistry.has(key);
```

Comment:
```suggestion
          return typeof key === 'string' && this._moduleRegistry.has(key);
```

Reply:
Unfortunately, `this` refers to ProxyHandler.

Reply:
@SimenB, How do you think, does it make sense to refer target props as fallback?
```typescript
get(target, key) {
          return typeof key === 'string'
            ? moduleRegistry.get(key) || target[key]
            : undefined;
        },
        has(target, key) {
          return typeof key === 'string'
            ? moduleRegistry.has(key) || !!target[key]
            : false;
        }
```

Reply:
can use arrow functions

Reply:
So be it.